### DISCUSSION TITLE
MAIN - time spent idling

### DISCUSSION THREAD

**Discussion by fcurow** [firemodels/fds#12954] [General]
To whom it may concern,

Looking at the cpu.csv file, it looks like the MAIN column dominates when compared to the other columns, but the MAIN is of the same order of magnitude for all cores in a parallel simulation, which suggests that all cores are relatively balanced in terms of time spent computing. What could be the cause of all cores idling on the same order of magnitude, but product a MAIN that is several times greater than any other column/activity? How sensitive should the comparison be when it comes to optimization between cores i.e. if there is a 5 or 10 % difference in MAIN time then it is fairly well balanced? Or should it be more like 1 to 2 % different? 

Best Regards,
Sal

**Comment by drjfloyd** [comment]
Not much happens in MAIN in terms of computational work. It is mostly just calls to initialization routines and then the main timestepping loop where computations are done elsewhere. The amount of work per process is going to be similar as all processes are taking the same number of timesteps. To see how well balanced things are look at the columns where computations are happening like DVIG, MASS, VELO, etc.

**Reply by fcurow** [reply]
Thank you.
The percent difference between the max and min of all the non-MAIN columns added between cores is less than 1%.

**Comment by mcgratta** [comment]
All of the FDS subroutines are timed explicitly, like DIVG, MASS, VELO, PRES, etc. The total CPU time minus the sum of all the explicitly timed routines is assigned to MAIN. If your MPI calculation is well-balanced, MAIN should be relatively low.

**Reply by fcurow** [reply]
Thank you.
I added up all the subroutines for each core and compared it to the MAIN, the ratio of MAIN to the sum of the other columns (not including Total T_USED) is about 9 times, this is consistent across all cores. Below is a comparison of the min and max values over all cores for each column. Does anything standout to you? The percent difference between the max and min of all the non-MAIN columns added between cores is less than 1%.

    	MAIN	DIVG	MASS	VELO	PRES	WALL	DUMP	PART	RADI	FIRE	COMM	EVAC	HVAC	
min	5E+05	4E+03	7E+02	3E+03	2E+03	4E+00	9E+01	3E+00	2E+03	6E+00	2E+04	0E+00	0E+00	
max	5E+05	2E+04	4E+03	1E+04	9E+03	8E+02	4E+02	4E+02	8E+03	4E+01	5E+04	0E+00	0E+00	


Rank	GEOM	VEGE	Total T_USED (s)
min	        0E+00	0E+00	6E+05
max         0E+00	0E+00	6E+05

**Reply by mcgratta** [reply]
Post the entire _cpu.csv file. I'm not sure I understand what you have written.

**Reply by fcurow** [reply]
[Test_cpu.csv](https://github.com/firemodels/fds/files/15357706/Test_cpu.csv)
[Test_cpu.csv](https://github.com/firemodels/fds/files/15357707/Test_cpu.csv)

**Reply by mcgratta** [reply]
![image](https://github.com/firemodels/fds/assets/11333911/c3fa3d57-7cbc-45bc-96b6-bf5b6510adab)
 There is something very wrong here. Are all meshes the same size?

**Reply by mcgratta** [reply]
What version of FDS is this?

**Reply by fcurow** [reply]
6.7.5

**Reply by mcgratta** [reply]
We're up to 6.9.1.  I suggest you try that. If you still get bad performance, post the input file and we'll run it on our computer.

**Comment by fcurow** [comment]
> ![image](https://private-user-images.githubusercontent.com/11333911/331720941-c3fa3d57-7cbc-45bc-96b6-bf5b6510adab.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTYyOTc5NTEsIm5iZiI6MTcxNjI5NzY1MSwicGF0aCI6Ii8xMTMzMzkxMS8zMzE3MjA5NDEtYzNmYTNkNTctN2NiYy00NWJjLTk2YjYtYmY1YjY1MTBhZGFiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTIxVDEzMjA1MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAyODQ5ZDc1MDdhYmE1OWJiZWM2ODMxMTdlZGQ1ZjI3ZjEzYjhlZjhlNjIzZDA2MWUzYTc0ZjNlZTc0Y2IzOTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.VFbUSJNh7olcJtSEzGTjCITWTw1WoyxGY-vnzyErReY) There is something very wrong here. Are all meshes the same size?

All meshes are about the same size, we tried to increase the number of meshes near the fire and there was no improvement in speed.

**Comment by fcurow** [comment]
We ran the case on another machine with 15 cores and it went much faster, the results you see for the ~70 core case are being run on HPC machines. So it seems like it has something to do with the computer/server set-up. We originally tried to run it on a network drive from the HPC computer, then thinking that might be the issue, we put it on the local drive (we avoided this because the local drive auto-deletes files after you log out), but there was no improvement. Any thoughts on what could cause this from a computer/server set-up perspective?

**Reply by drjfloyd** [reply]
A couple of things to look at on the HPC (this may take some support from your local IT / sysadmin):

Are you oversubscribing cores (running more than one process on a core) on the HPC when a job is submitted? 
Is hyperthreading enabled on the HPC? FDS is memory intensive and using virtualized cores can bottleneck the memory bus.

In the FDS github repository there are two sets of scaling tests:
https://github.com/firemodels/fds/tree/master/Validation/MPI_Scaling_Tests/FDS_Input_Files

These are discussed in the User's Guide.  Running these may give some insights on your particular system that might help your IT pinpoint issues.

**Reply by fcurow** [reply]
Thank you, it is also worth mentioning that:
1) the model was being run on an HPC Azure instance 
2) the cores were locked to 2.59Ghz

**Reply by drjfloyd** [reply]
I have little experience running FDS on Azure, but I know that both Azure and AWS have a lot of options in terms of hardware (type of CPU, core count, memory, networking, etc.) for running jobs. You may have picked a hardware option not well suited for running FDS. For example, if your 70 core case was some option using multiple machines without a fast network connection.

**Reply by fcurow** [reply]
Thank you, is there a recommended configuration with Azure?

**Reply by drjfloyd** [reply]
I would not be able to answer that. If you have an instutional account your account manager may have access to support resources. You could ask this question of the general FDS community, but in that case I would make a new discussion with a relevant title.

**Comment by fcurow** [comment]
The engineer that was running the simulation forgot to change the OPEN_MP to 1, it was set to 3, the simulation is running a lot faster now. Sorry for the distraction. Thank you.
